### PR TITLE
Merge Entity Inspector and Asset Browser Inspector

### LINKED ISSUE #12876: 27 - Asset Browser - Context aware inspector (Merge inspectors)
User Story: As a user I would like one inspector instead of managing several inspectors within my workflows. This new inspector experience should be context aware and always show what is in my selection state. This means if I'm clicking on asset in my asset browser it should show asset related details. If I'm clicking on entities or prefabs it should show the current entity inspector. Making the inspector a general inspector instead of separate inspectors for different experiences. Then we would drop all the secondary naming conventions and just call it the "Inspector".

 

I've also hear from other LY team such as animation and large world teams about utilizing a global that can cover more of everyone experience. Example the animation inspector can be use in the main inspector. Or landscape canvas can inspect nodes in the main window.   Simplifying the approach for the user of having to keep track of multiple open panels

 

image

### PR DESCRIPTION
**Pull request by dtamkin1**
## What does this PR do?

Combines the Entity Inspector and the Asset Browser Inspector into a stacked widget that displays the appropriate inspector based on entity or asset selection.

- Rename from "Entity Inspector" to "Inspector", update AutomatedTesting and variable references
- Allows users to browse through folders and directories without switching the entity inspector
- Allows users to drag-and-drop assets into component fields
- Allows users to pin an entity inspector and have it side by side with the new inspector

This is the first of many changes to creating a more robust Asset Browser Inspector, see https://github.com/orgs/o3de/projects/16 for more details.

Fixes #12876 

Editor_cv4VOmUOXy

## How was this PR tested?

Locally, opening and closing the inspector, enabled and disabled the DPE to make sure functionality remains the same

### CODE REVIEW

**Review by cgalvan — COMMENTED** [review]
It sounds like UX is actually requesting a new tool called "Inspector" that would be a combination of the Entity Inspector and Asset Browser Inspector, as folding the Asset Browser Inspector into the Entity Inspector.

Should be straightforward to do something similar to what you have here with the stacked widget but just in a new tool. @rainbj  ?

**Comment by dtamkin1** [comment]
> It sounds like UX is actually requesting a new tool called "Inspector" that would be a combination of the Entity Inspector and Asset Browser Inspector, as folding the Asset Browser Inspector into the Entity Inspector.
> 
> 
> 
> Should be straightforward to do something similar to what you have here with the stacked widget but just in a new tool. @rainbj  ?

Yeah I wasn't sure about this either so I pinged Josh about it, but I think I could just change the name of the Entity Inspector to Inspector if that's the case though right?

**Comment by cgalvan** [comment]
> > It sounds like UX is actually requesting a new tool called "Inspector" that would be a combination of the Entity Inspector and Asset Browser Inspector, as folding the Asset Browser Inspector into the Entity Inspector.
> > Should be straightforward to do something similar to what you have here with the stacked widget but just in a new tool. @rainbj  ?
> 
> Yeah I wasn't sure about this either so I pinged Josh about it, but I think I could just change the name of the Entity Inspector to Inspector if that's the case though right?

I think the problem workflow-wise would be that some people might not want their Entity Inspector to change if they are selecting things in the Asset Browser, which is echo'd by @AMZN-stankowi comment on #12876

**Comment by rainbj** [comment]
Hi @cgalvan and @AMZN-stankowi, trying to address your concern from above. So just an FYI, this pattern is used in all game engine except ours.  Unity, Unreal, Godot have a context-aware selection model. We have also received feedback from users directly around asset selection and how the best content switch would be when selecting an asset. In most of those cases, users preferred a method to "other game engines and how they use selection state". So yes this will be a shift in mindset for some. However...

For the specific use case that @AMZN-stankowi mentioned above. Being able to see both at the same time. My assumption is the feature we built a couple of years ago. "Pinning an inspector" would allow us to do both. Keep state of the pinned object open as well as explore content.

User clicks on pin


New window open. User can now click on asset.

**Review by rainbj — COMMENTED** [review]
UX approved. However, I don't feel comfortable approving code as this is not my area of expertise

**Comment by lsemp3d** [comment]
> > > It sounds like UX is actually requesting a new tool called "Inspector" that would be a combination of the Entity Inspector and Asset Browser Inspector, as folding the Asset Browser Inspector into the Entity Inspector.
> > > Should be straightforward to do something similar to what you have here with the stacked widget but just in a new tool. @rainbj  ?
> > 
> > 
> > Yeah I wasn't sure about this either so I pinged Josh about it, but I think I could just change the name of the Entity Inspector to Inspector if that's the case though right?
> 
> I think the problem workflow-wise would be that some people might not want their Entity Inspector to change if they are selecting things in the Asset Browser, which is echo'd by @AMZN-stankowi comment on #12876

Out of curiosity, does it add complexity or confusion if the behavior is a user preference?

File: Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp [hunk 1]
```diff
@@ -779,6 +779,7 @@ void SandboxIntegrationManager::OpenPinnedInspector(const AzToolsFramework::Enti
         if (editor && editor->GetPropertyEditor())
         {
             editor->GetPropertyEditor()->SetOverrideEntityIds(entities);
+            editor->SetPinnedInspector();
```
Comment (cgalvan) [inline Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp:782]: We can already deduce if the inspector is pinned or not by checking if `m_overrideSelectedEntityIds` is empty or not. So we don't need to add a new `m_isInspectorPinned` variable to track.

To make it more readable though you could add a new method like `bool InspectorIsPinned() const` that you could use and replace the handful of instances in `EntityPropertyEditor` that are currently checking `m_overrideSelectedEntityIds.empty()`

**Review by cgalvan — CHANGES_REQUESTED** [review]
Also if we change the name to just `Inspector` we'll need to update these references in automated tests as well:

```Search "Entity Inspector" (49 hits in 16 files of 1120 searched)
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\Atom\tests\hydra_AtomEditorComponents_MeshAdded.py (3 hits)
	Line 298:         # Make sure the Entity Inspector is open and trigger the "Add Material Component" button
	Line 299:         general.open_pane("Entity Inspector")
	Line 301:         entity_inspector = editor_window.findChild(QtWidgets.QDockWidget, "Entity Inspector")
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\Atom\tests\__pycache__\hydra_AtomEditorComponents_MeshAdded.cpython-37.pyc (1 hit)
	Line 70: base_emptyé
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\editor\EditorScripts\AssetPicker_UI_UX.py (3 hits)
	Line  30:         3) Access Entity Inspector
	Line 225:         # 3) Access Entity Inspector
	Line 227:         entity_inspector = editor_window.findChild(QtWidgets.QDockWidget, "Entity Inspector")
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\editor\EditorScripts\BasicEditorWorkflows_LevelEntityComponentCRUD.py (1 hit)
	Line 113:         # 2) Delete existing entities, and create and manipulate new entities via Entity Inspector
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\editor\EditorScripts\ComponentCRUD_Add_Delete_Components.py (3 hits)
	Line  45:         3) Components can be selected in the Entity Inspector.
	Line 112:         # Give the Entity Inspector time to fully create its contents
	Line 117:         entity_inspector = editor_window.findChild(QtWidgets.QDockWidget, "Entity Inspector")
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\editor\EditorScripts\Docking_BasicDockedTools.py (14 hits)
	Line  21:         "Entity Inspector works when docked, Entity name changed",
	Line  49:             2.3) Select the Entity Inspector in the floating window.
	Line  50:             2.4) Change the name of the selected Entity via the Entity Inspector.
	Line  71:         # Make sure the Entity Outliner, Entity Inspector and Console tools are open
	Line  73:         general.open_pane("Entity Inspector")
	Line  93:         # Dock the Entity Inspector tabbed with the floating Entity Outliner
	Line  94:         entity_inspector = editor_window.findChild(QtWidgets.QDockWidget, "Entity Inspector")
	Line  97:         # We need to grab a new reference to the Entity Inspector QDockWidget because when it gets moved
	Line  99:         entity_inspector = editor_window.findChild(QtWidgets.QDockWidget, "Entity Inspector")
	Line 101:         # Dock the Console tabbed with the floating Entity Inspector
	Line 107:             entity_inspector = editor_window.findChild(QtWidgets.QDockWidget, "Entity Inspector")
	Line 113:             Report.info(f"Entity Inspector parent = {entity_inspector_parent}, Entity Outliner parent = "
	Line 123:         entity_inspector = editor_window.findChild(QtWidgets.QDockWidget, "Entity Inspector")
	Line 132:         # 2.3,4) Change the name of the selected Entity via the Entity Inspector.
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\editor\EditorScripts\__pycache__\AssetPicker_UI_UX.cpython-310.pyc (2 hits)
	Line  42:         3) Access Entity Inspector
	Line 116: TestEntityZMeshzEntity InspectorZm_componentListContentsrB
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\editor\EditorScripts\__pycache__\Docking_BasicDockedTools.cpython-310.pyc (5 hits)
	Line 10: c
	Line 33:             2.3) Select the Entity Inspector in the floating window.
	Line 34:             2.4) Change the name of the selected Entity via the Entity Inspector.
	Line 40:         r
	Line 41: renderOverlayc
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\EditorPythonBindings\tests\Editor_PySide_Example_Works.py (1 hit)
	Line 36:         general.open_pane("Entity Inspector")
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\EditorPythonTestTools\editor_python_test_tools\pyside_component_utils.py (4 hits)
	Line 15:     visible in the Entity Inspector.
	Line 16:     Works by inspecting the list of child widgets for the Entity from the Entity Inspector, looking for the right type
	Line 30:     entity_inspector = editor_window.findChild(PySide2.QtWidgets.QDockWidget, 'Entity Inspector')
	Line 31:     assert entity_inspector, 'Entity Inspector widget is not valid.'
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\largeworlds\landscape_canvas\EditorScripts\ComponentUpdates_UpdateGraph.py (2 hits)
	Line  75:         5. Delete the Vegetation Distribution Filter component from the BushSpawner entity via Entity Inspector
	Line  78:         7. Add Vegetation Altitude Filter to the BushSpawner entity through Entity Inspector
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\largeworlds\landscape_canvas\EditorScripts\SlotConnections_UpdateComponentReferences.py (2 hits)
	Line  45:     A reference created through slot connections in Landscape Canvas is reflected in the Entity Inspector.
	Line  51:      4) Component references are verified via Entity Inspector
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\largeworlds\landscape_canvas\EditorScripts\__pycache__\ComponentUpdates_UpdateGraph.cpython-310.pyc (2 hits)
	Line 32:         5. Delete the Vegetation Distribution Filter component from the BushSpawner entity via Entity Inspector
	Line 35:         7. Add Vegetation Altitude Filter to the BushSpawner entity through Entity Inspector
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\largeworlds\landscape_canvas\EditorScripts\__pycache__\ComponentUpdates_UpdateGraph.cpython-37.pyc (2 hits)
	Line 32:         5. Delete the Vegetation Distribution Filter component from the BushSpawner entity via Entity Inspector
	Line 35:         7. Add Vegetation Altitude Filter to the BushSpawner entity through Entity Inspector
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\largeworlds\landscape_canvas\EditorScripts\__pycache__\SlotConnections_UpdateComponentReferences.cpython-310.pyc (2 hits)
	Line 26:     A reference created through slot connections in Landscape Canvas is reflected in the Entity Inspector.
	Line 32:      4) Component references are verified via Entity Inspector
  D:\github\o3de\AutomatedTesting\Gem\PythonTests\largeworlds\landscape_canvas\EditorScripts\__pycache__\SlotConnections_UpdateComponentReferences.cpython-37.pyc (2 hits)
	Line 25:     A reference created through slot connections in Landscape Canvas is reflected in the Entity Inspector.
	Line 31:      4) Component references are verified via Entity Inspector
```

File: Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp (hunk 1 above)
Comment (rainbj) [reply Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp:782]: Great work guys. This is a big deal and I'm loving this. I imagine that when we start building out all the asset settings. We might also want to add the pin into the other side of the inspector. Then you can "pin" any kind of inspector. But just some long term thinking

Comment (dtamkin1) [reply Code/Editor/Plugins/ComponentEntityEditorPlugin/SandboxIntegration.cpp]: It turns out the `EntityPropertyEditor` already has a method called `IsLockedToSpecificEntities()` that returns true if `m_overrideSelectedEntityIds` is not empty, so I replaced it with that

**Review by cgalvan — APPROVED** [review]

**Review by lsemp3d — APPROVED** [review]

**Pull request merged**